### PR TITLE
PostgreSQL: prepare for pg-1.1

### LINKED ISSUE #33185: Next version of PG will deprecate calling async_exec with params
async_exec should no longer be called with params and instead async_exec_params needs to be used. This pretty enormous deprecation will hit next time PG gem is released so we really got to get it sorted asap.

`/home/sam/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/postgresql_adapter.rb:603: warning: forwarding async_exec to async_exec_params and send_query to send_query_params is deprecated`

(anyone using prepared statements will be impacted)

Example gist:

https://gist.github.com/2f3ed36aa4b7ff0875e035634eff3ff2

cc @larskanis

### PR DESCRIPTION
**Pull request by larskanis**
pg-1.1 is planned to deprecate some behavior. I (tried to) discussed this [here](https://bitbucket.org/ged/ruby-pg/issues/277/deprecation-warnings-for-query-methods). It currently warns at every method invocation, but I plan to change it to warm only once and to add an environment variable to disable the warning entirely. I'm also open to disable the most prominent warning about async_exec vs. async_exec_params until the next rails-5.x releases are out.

Fixes #33185

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/postgresql/oid/array.rb
Comment (SamSaffron) [inline activerecord/lib/active_record/connection_adapters/postgresql/oid/array.rb]: the inline rescue there is a bit worrying, why is it needed?

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb
Comment (SamSaffron) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: Would exec_params be safe as a 5.2 backport? (would we need to backport to 5.1/5.0/4.2 as well?)

Comment (SamSaffron) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: also I find the semantics a bit confusing cause I would expect exec_params to work like exec and async_exec_params to work like async_exec ... but it appears that exec_params is always async.

File: activerecord/lib/active_record/connection_adapters/postgresql/oid/array.rb
Comment (larskanis) [reply activerecord/lib/active_record/connection_adapters/postgresql/oid/array.rb]: This method receives user inputs and cast it to Array. They can be malformed and that will be raised in pg-2.0. See [this commit](https://bitbucket.org/ged/ruby-pg/commits/1b081326b346368e70c9c03ee7080e28d6b3a3dc). The rescue is not there for values retrieved from the database.

**Comment by larskanis** [comment]
This should be backported to all rails-5.x. rails-4.x is not (yet) compatible to pg-1.x so backporting makes no sense there.

**Comment by matthewd** [comment]
> I (tried to) discussed this here

Ah sorry, that's not me.

File: activerecord/lib/active_record/connection_adapters/postgresql/oid/array.rb
Comment (SamSaffron) [reply activerecord/lib/active_record/connection_adapters/postgresql/oid/array.rb]: maybe we explicitly rescue TypeError then with a tiny comment about it?

```
value = begin
    @pg_decoder.decode(value)
  rescue TypeError
     # malformed array string is treated as [], will raise in PG 2.0 gem
     # this keeps a consistent implementation 
     []
   end
```

I do wonder if we should just let this error float back to the user cause they are sending in something PG can not handle so its surprising to have input eaten.

**Comment by matthewd** [comment]
Yeah I think my suggestion (and I'm aware this is coming from a maintainer of a very deprecation-happy project) would be to keep the old names working at least for one more release series, as they don't seem to be "in the way" / occupying names you want to reuse with different semantics.

I think we're currently semi-unofficially still mostly maintaining 5.0, so a full 5.x backport isn't impossible... but it does seem a bit pointless: much as 4.2 doesn't support 1.x, 5.x doesn't support pg 2.x -- so it's a warning about a change that code will never encounter. :confused:

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb
Comment (larskanis) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: > Would exec_params be safe as a 5.2 backport?

Yes, `exec_params` has been there since pg-0.15.0. And starting with pg-1.1.0 it will be a alias for `async_exec_params`.

In pg < 1.1 `exec_params` did what `sync_exec_params` will do in pg-1.1. `sync_exec_params` differed from `async_exec_params`, in that it doesn't stop at signals. It delayes signal processing until the SQL statement has been finished.

Using `exec_params` instead of `async_exec` is therefore kind of regression regarding signals. But there are still other pg methods in use which delay signal processing equally and it is only valid when running with pg < 1.1, so that I don't think it's actually a problem. I asked to get the other methods fixed in PR #32820, but then we came to the conclusion, that it's better to alias all query methods to the corresponding `async_foo` method in pg-1.1 instead.

**Comment by SamSaffron** [comment]
Lars, thinking about this, perhaps we simply strip the warning for now from 1.1 (and re-introduce with 2.0) and then "eat" the behavior change with malformed arrays on upgrade of PG

The trouble of getting this into Rails and backported is that we will be that we are 100% blocked on releasing PG 1.1 till a new gem is cut for 5.X, this usually only happens when there is a security flaw found. It could be months.

I think for Rails 6.0 we should change it so malformed arrays are malformed arrays and raise properly, nuking that weird test. 

@matthewd  @sgrif / @tenderlove  thoughts?

**Comment by y-yagi** [comment]
It seems that `pg` 1.1.0 has been released. And when use it, warning about `async_exec` is showed. 

```
warning: forwarding async_exec to async_exec_params and send_query to send_query_params is deprecated 
```

I'd like to fix that, but what is the status of this PR now?
Is it better to deal with another PR?

File: activerecord/lib/active_record/connection_adapters/postgresql/oid/array.rb
Comment (larskanis) [inline activerecord/lib/active_record/connection_adapters/postgresql/oid/array.rb]: I made it [a configuration option](https://github.com/ged/ruby-pg/commit/ee1e19cca83ac0b5d01b7776997964c300445e24) whether a TypeError is raised or whether a partial array is returned. The latter one is the default, so that the second commit of this PR is not strictly necessary for pg-1.1.0.

**Comment by y-yagi** [comment]
Thank you for the explanation! I understood.

**Review by y-yagi — APPROVED** [review]

**Review by plentz — APPROVED** [review]

**Comment by y-yagi** [comment]
@larskanis Thanks!

**Pull request merged**
